tests/pipeline/test_validation.py

```diff
@@ -0,0 +1,166 @@
+from pydantic import ValidationError
+import pytest
+
+from chatsky.pipeline import (
+    Pipeline,
+    Service,
+    ServiceGroup,
+    Actor,
+    ServiceRuntimeInfo,
+    BeforeHandler,
+)
+from chatsky.script import Context
+from chatsky.utils.testing import TOY_SCRIPT, TOY_SCRIPT_KWARGS
+
+
+# Looks overly long, we only need one function anyway.
+class UserFunctionSamples:
+    """
+    This class contains various examples of user functions along with their signatures.
+    """
+
+    @staticmethod
+    def correct_service_function_1(_: Context):
+        pass
+
+    @staticmethod
+    def correct_service_function_2(_: Context, __: Pipeline):
+        pass
+
+    @staticmethod
+    def correct_service_function_3(_: Context, __: Pipeline, ___: ServiceRuntimeInfo):
+        pass
+
+
+# Could make a test for returning an awaitable from a ServiceFunction, ExtraHandlerFunction
+class TestServiceValidation:
+    def test_model_validator(self):
+        with pytest.raises(ValidationError) as e:
+            # Can't pass a list to handler, it has to be a single function
+            Service(handler=[UserFunctionSamples.correct_service_function_2])
+            assert e
+        with pytest.raises(ValidationError) as e:
+            # 'handler' is a mandatory field
+            Service(before_handler=UserFunctionSamples.correct_service_function_2)
```

Comment:
Tests don't need to be so thorough.
In this case we rely on Pydantic and developers/reviewers (to ensure correct field definition) to make sure that a service cannot be created without passing a handler.